src/ObjectPool/src/DisposableObjectPool.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Extensions.ObjectPool
+{
+    public class DisposableObjectPool<T> : DefaultObjectPool<T>, IDisposable where T : class
+    {
+        private volatile bool _isDisposed;
+
+        public DisposableObjectPool(IPooledObjectPolicy<T> policy)
+            : base(policy)
+        {
+        }
+
+        public DisposableObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
+            : base(policy, maximumRetained)
+        {
+        }
+
+        public override T Get()
+        {
+            if (_isDisposed)
+            {
+                ThrowObjectDisposedException();
+            }
+
+            return base.Get();
+
+            void ThrowObjectDisposedException()
+            {
+                throw new ObjectDisposedException(this.GetType().Name);
+            }
+        }
+
+        public override void Return(T obj)
+        {
```

Comment:
Should this have a null-check?

Reply:
I don't see any need for the null-check.

If `obj != null` everything is fine.
if `obj == null` the return-code will handle it, for the disposed-case there is `item is IDisposable`, so if null the branch won't be taken.